File: Leetcode/C++/TwoPointer/isPalindrome.cpp
/*
Solution: 双指针，时间复杂度O(n),空间复杂度O(1)
- 定义两个指针p,q分别指向字符串开始于结尾；
- 用isalnum判断是否为字母数字，否则指针右移或左移；
- 用toupper来判断大小写字母；
*/

class Solution {
public:
	bool isPalindrome(string s) {
		if (s == "")
			return true;

		int p = 0, q = s.size() - 1;
        while(p < q){
            while(!isalnum(s[p]) && p < q)   p++;
            while(!isalnum(s[q]) && p < q)   q--;
            if(toupper(s[p]) != toupper(s[q]))  
                return false;
            else{
                p++; q--;
            }
        }
        return true;
        
	}
};

File: Leetcode高频题/isPalindrome.py
'''
@lsy 2019.11.26

双指针
注意：转换为小写islower()；判断是否为数字和字母isalnum()
'''
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if not s:
            return True
        
        s = s.lower()
        
        i, j = 0, len(s) - 1
        while i < j:
            while i < j and not s[i].isalnum():
                i += 1
            while i < j and not s[j].isalnum():
                j -= 1
            
            if s[i] != s[j]:
                return False
            
            i, j = i + 1, j - 1
        
        return True
                